Makefile

```diff
@@ -126,6 +126,27 @@ push:
 		$(DOCKER) push $(REGISTRY)/$(TARGET):latest; \
 	fi
 
+travis_image_push: container
+
+	if [ $(GIT_TAG) ] && [ $(GIT_TAG)x = $(TRAVIS_TAG)x ] || [ $(TRAVIS_BRANCH)x = masterx ]; \
```

Comment:
`TRAVIS_BRANCH` is always going to be set here as for builds triggered by a pull request this is the name of the branch targeted by the pull request i.e. `master`. The old script even had a bug here as can be seen in a recent travis run https://travis-ci.org/kubernetes-sigs/federation-v2/builds/527879538#L3120. I suggest using the `TRAVIS_PULL_REQUEST` environment variable to determine whether `docker login` is necessary as it is:

> The pull request number if the current job is a pull request, “false” if it’s not a pull request.

Then you can move this logic within the `push` target and just add the `master` check to it for the `canary` `tag` and `push` commands.

Also, is the trailing `x` necessary? Presumably we should be using `bash`.

Reply:
Thank you @font . Updated the PR, please help take a look.